File: ttt.py
#!/usr/bin/env python
'''Tic tac toe in python, Minimax with alpha-beta pruning.'''
import sys
import random
import getopt

# Board: array of 9 int, positionally numbered like this:
# 0  1  2
# 3  4  5
# 6  7  8

# Well-known board positions
WINNING_TRIADS = ((0, 1, 2), (3, 4, 5), (6, 7, 8), (0, 3, 6), (1, 4, 7),
    (2, 5, 8), (0, 4, 8), (2, 4, 6))
PRINTING_TRIADS = ((0, 1, 2), (3, 4, 5), (6, 7, 8))
# The order in which slots get checked for absence of a player's token:
SLOTS = (0, 1, 2, 3, 4, 5, 6, 7, 8)

# Internal-use values.  Chosen so that the "winner" of a finished
# game has an appropriate value, as X minimizes and O maximizes
# the board's value (function board_valuation() defines "value")
# Internally, the computer always plays Os, even though the markers[]
# array can change based on -r command line flag.
X_token = -1
Open_token = 0
O_token = 1

# Strings for output: player's markers, phrase for end-of-game
MARKERS = ['_', 'O', 'X']
END_PHRASE = ('draw', 'win', 'loss')

HUMAN = 1
COMPUTER = 0

def board_valuation(board, player, next_player, alpha, beta):
    '''Dynamic and static evaluation of board position.'''
    # Static evaluation - value for next_player
    wnnr = winner(board)
    if wnnr != Open_token:
        # Not a draw or a move left: someone won
        return wnnr
    elif not legal_move_left(board):
        # Draw - no moves left
        return 0 # Cat
    # If flow-of-control gets here, no winner yet, not a draw.
    # Check all legal moves for "player"
    for move in SLOTS:
        if board[move] == Open_token:
            board[move] = player
            val = board_valuation(board, next_player, player, alpha, beta)
            board[move] = Open_token
            if player == O_token:  # Maximizing player
                if val > alpha:
                    alpha = val
                if alpha >= beta:
                    return beta
            else:  # X_token player, minimizing
                if val < beta:
                    beta = val
                if beta <= alpha:
                    return alpha
    if player == O_token:
        retval = alpha
    else:
        retval = beta
    return retval

def print_board(board):
    '''Print the board in human-readable format.
       Called with current board (array of 9 ints).
    '''
    for row in PRINTING_TRIADS:
        for hole in row:
            print MARKERS[board[hole]],
        print

def legal_move_left(board):
    ''' Returns True if a legal move remains, False if not. '''
    for slot in SLOTS:
        if board[slot] == Open_token:
            return True
    return False

def winner(board):
    ''' Returns -1 if X wins, 1 if O wins, 0 for a cat game,
        0 for an unfinished game.
        Returns the first "win" it finds, so check after each move.
        Note that clever choices of X_token, O_token, Open_token
        make this work better.
    '''
    for triad in WINNING_TRIADS:
        triad_sum = board[triad[0]] + board[triad[1]] + board[triad[2]]
        if triad_sum == 3 or triad_sum == -3:
            return board[triad[0]]  # Take advantage of "_token" values
    return 0

def determine_move(board):
    ''' Determine Os next move. Check that a legal move remains before calling.
        Randomly picks a single move from any group of moves with the same value.
    '''
    best_val = -2  # 1 less than min of O_token, X_token
    my_moves = []
    for move in SLOTS:
        if board[move] == Open_token:
            board[move] = O_token
            val = board_valuation(board, X_token, O_token, -2, 2)
            board[move] = Open_token
            print "My move", move, "causes a", END_PHRASE[val]
            if val > best_val:
                best_val = val
                my_moves = [move]
            if val == best_val:
                my_moves.append(move)
    return random.choice(my_moves)

def recv_human_move(board):
    ''' Encapsulate human's input reception and validation.
        Call with current board configuration. Returns
        an int of value 0..8, the Human's move.
    '''
    looping = True
    while looping:
        try:
            inp = input("Your move: ")
            yrmv = int(inp)
            if 0 <= yrmv <= 8:
                if board[yrmv] == Open_token:
                    looping = False
                else:
                    print "Spot already filled."
            else:
                print "Bad move, no donut."

        except EOFError:
            print
            sys.exit(0)
        except NameError:
            print "Not 0-9, try again."
        except SyntaxError:
            print "Not 0-9, try again."

        if looping:
            print_board(board)

    return yrmv

def usage(progname):
    '''Call with name of program, to explain its usage.'''
    print progname + ": Tic Tac Toe in python"
    print "Usage:", progname, "[-h] [-c] [-r] [-x] [-X]"
    print "Flags:"
    print "-x, -X:   print this usage message, then exit."
    print "-h:  human goes first (default)"
    print "-c:  computer goes first"
    print "-r:  computer is X, human is O"
    print "The computer O and the human plays X by default."

def main():
    '''Call without arguments from __main__ context.'''
    try:
        opts, args = getopt.getopt(sys.argv[1:], "chrxX",
            ["human", "computer", "help"])
    except getopt.GetoptError:
        # print help information and exit:
        usage(sys.argv[0])
        sys.exit(2)

    next_move = HUMAN # Human goes first by default

    for opt, arg in opts:
        if opt == "-h":
            next_move = HUMAN
        if opt == "-c":
            next_move = COMPUTER
        if opt == "-r":
            MARKERS[-1] = 'O'
            MARKERS[1] = 'X'
        if opt in ("-x", "-X", "--help"):
            usage(sys.argv[0])
            sys.exit(1)

    # Initial state of board: all open spots.
    board = [Open_token, Open_token, Open_token, Open_token, Open_token,
        Open_token, Open_token, Open_token, Open_token]

    # State machine to decide who goes next, and when the game ends.
    # This allows letting computer or human go first.
    while legal_move_left(board) and winner(board) == Open_token:
        print
        print_board(board)

        if next_move == HUMAN and legal_move_left(board):
            humanmv = recv_human_move(board)
            board[humanmv] = X_token
            next_move = COMPUTER

        if next_move == COMPUTER and legal_move_left(board):
            mymv = determine_move(board)
            print "I choose", mymv
            board[mymv] = O_token 
            next_move = HUMAN

    print_board(board)
    # Final board state/winner and congratulatory output.
    try:
        # "You won" should never appear on output: the program
        # should always at least draw.
        print ["Cat got the game", "I won", "You won"][winner(board)]
    except IndexError:
        print "Really bad error, winner is", winner(board)

    sys.exit(0)
#-------
if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print
        sys.exit(1)

File: ttt.c
#include <stdio.h>
#include <stdlib.h>
/* Board is an 8-element array of ints, organize 0-based indexes
 * like this:
 *
 *  0   1   2
 *  3   4   5
 *  6   7   8
 *
 */

int winning_triads[8][3] = {
	{0,1,2}, {3,4,5}, {6,7,8},
	{0,3,6}, {1,4,7}, {2,5,8},
	{0,4,8}, {2,4,6}
};

int  find_winner(int board[]);     /* if computer wins 1, if humn -1, 0 no win */
int  move_remaining(int board[]);  /* 1 if a legal move remains, 0 otherwise */
int  computer_moves(int board[]);  /* determine computer's next move */
int  get_human_move(int board[]);  /* get an int between 0 and 8 inclusive */
void print_board(int board[]);

int
main(int ac, char **av)
{
	int goes_first = -1;
	int board[9];
	int winner = 0;
	int new_move;
	int i;

	if (ac > 1)
	{
		if (av[1][0] == 'h' || av[1][0] == 'H')
			goes_first = -1;
		else if (av[1][0] == 'c' || av[1][0] == 'C')
			goes_first = 1;
	}

	printf("0   1   2\n3   4   5\n6   7   8\n");

	for (i = 0; i < 9; ++i)
		board[i] = 0;

	if (goes_first == -1)
		printf("\nHuman plays first, X, computer plays 2nd, O\n");
	else
		printf("\nComputer plays first, 0, human plays 2nd, X\n");

	/* Duff-like device. */
	switch (goes_first)
	{
	do {

	case -1:
		new_move = get_human_move(board);  /* human's next move */
		board[new_move] = -1; /* Human is minimizer */
		winner = find_winner(board);
		if (winner) break;

	case 1:
		new_move = computer_moves(board);
		board[new_move] = 1; /* Computer maximises */
		winner = find_winner(board);
		if (winner) break;

		print_board(board);

	} while (0 == winner && move_remaining(board));

	}

	switch (winner)
	{
	case -1:  printf("\nHuman wins.\n"); break;
	case 1:   printf("\nComputer wins\n"); break;
	default:  printf("\nCat got the game\n"); break;
	}

	print_board(board);

	return 0;
}

int
move_remaining(int board[])
{
	int i, legal_move_left = 0;
	for (i = 0; i < 9; ++i)
	{
		if (board[i] == 0)
		{
			legal_move_left = 1;
			break;
		}
	}
	return legal_move_left;
}

int
find_winner(int board[])
{
	int i;
	int winner = 0;

	for (i = 0; i < 8; ++i)
	{
		int sum;
		int *triad = winning_triads[i];

		sum = board[triad[0]] + board[triad[1]] + board[triad[2]];

		if (sum == 3)
			winner = 1;
		else if (sum == -3)
			winner = -1;
	}

	return winner;
}

int corners[] = {0, 2, 6, 8};
int sides[] = {1, 3, 5, 7};

int
computer_moves(int board[])
{
	int i;

	/* fill in any wins */
	for (i = 0; i < 8; ++i)
	{
		int sum;
		int *triad = winning_triads[i];

		sum = board[triad[0]] + board[triad[1]] + board[triad[2]];

		if (sum == 2)
		{
			int j;
			for (j = 0; j < 3; ++j)
			{
				if (board[triad[j]] == 0)
					return triad[j];
			}
		}
	}

	/* block any 2-in-a row */
	for (i = 0; i < 8; ++i)
	{
		int sum;
		int *triad = winning_triads[i];

		sum = board[triad[0]] + board[triad[1]] + board[triad[2]];

		if (sum == -2)
		{
			int j;
			for (j = 0; j < 3; ++j)
			{
				if (board[triad[j]] == 0)
					return triad[j];
			}
		}
	}

	/* Move to center if possible */
	if (board[4] == 0)
		return 4;

	/* Move to a corner if possible */
	for (i = 0; i < 3; ++i)
		if (board[corners[i]] == 0)
			return corners[i];

	/* Move to a side if possible */
	for (i = 0; i < 3; ++i)
		if (board[sides[i]] == 0)
			return sides[i];

	return -1;
}

void
print_board(int board[])
{
		printf("%c   %c   %c\n%c   %c   %c\n%c   %c   %c\n",
			(board[0] == -1? 'X': (board[0] == 0? '_': 'O')),
			(board[1] == -1? 'X': (board[1] == 0? '_': 'O')),
			(board[2] == -1? 'X': (board[2] == 0? '_': 'O')),
			(board[3] == -1? 'X': (board[3] == 0? '_': 'O')),
			(board[4] == -1? 'X': (board[4] == 0? '_': 'O')),
			(board[5] == -1? 'X': (board[5] == 0? '_': 'O')),
			(board[6] == -1? 'X': (board[6] == 0? '_': 'O')),
			(board[7] == -1? 'X': (board[7] == 0? '_': 'O')),
			(board[8] == -1? 'X': (board[8] == 0? '_': 'O'))
		);
}

int
get_human_move(int board[])
{
	char buffer[16];
	int new_move = -1;  /* human's next move */

	while (new_move == -1)
	{
		int candidate;
		printf("Your move > ");
		fgets(buffer, sizeof(buffer), stdin);
		printf("\n");
		candidate = strtol(buffer, NULL, 10);
		if (0 <= candidate && 9 > candidate && board[candidate] == 0)
		{
			new_move = candidate;
			break;
		} else {
			if (0 > candidate || 9 <= candidate)
				fprintf(stderr, "Valid move between 0 and 8\n");
			if (board[candidate] != 0)
				fprintf(stderr, "That spot is taken\n");
		}
	}

	return new_move;
}
